Add fix qtpie/reaxff

navlalli opened the pull request:
**Summary**

`fix qtpie/reaxff` performs charge equilibration using the [QTPIE](https://doi.org/10.1016/j.cplett.2007.02.065) charge equilibration scheme when using `pair_style reaxff`. This fix can successfully shield long range charge transfer commonly seen with the QEq charge equilibration scheme and allows for periodic boundaries in the direction of an external electric field.

**Related Issue(s)**

Closes #4226 

<!--If this addresses an open GitHub issue for this project, please mention the issue number here, and describe the relation. Use the phrases `fixes #221` or `closes #135`, when you want an issue to be automatically closed when the pull request is merged-->

**Author(s)**

Navraj S Lalli (Imperial College London)
Efstratios M Kritikos (California Institute of Technology)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Changes in this pull request will not break backward compatibility.

**Implementation Notes**

This fix is implemented by adding fix_qtpie_reaxff.cpp and fix_qtpie_reaxff.h to the REAXFF package.

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

navlalli committed:
Create fix qtpie/reaxff

navlalli committed:
Allow for reading of Gaussian exponents from file

navlalli committed:
Add functionality for calculating chi_eff

navlalli committed:
Rename variables in calc_chi_eff()

navlalli committed:
Update calculation of chi_eff

chi_eff can now be calculated when atom or equal style variables are
used for the electric field, under the restriction that the electric
field is applied to all atoms.

navlalli committed:
Remove chi_field

navlalli committed:
Replace qeq with qtpie

navlalli committed:
Remove dual as a possible keyword argument

navlalli committed:
Remove virtual keyword

navlalli committed:
Add fix_qtpie_reaxff

navlalli committed:
Update contributing authors

navlalli committed:
Remove unused include

navlalli committed:
Rename variable for sum of local and ghost atoms

navlalli committed:
Rename misleading variable name

navlalli committed:
Correct calculation of cut off distance

navlalli committed:
Improve exceptions in reading of gauss file

navlalli committed:
Update with changes made to fix_qeq_reaxff.cpp

akohlmey commented:
@navlalli please make sure you click on "Ready for review" when you want people to look at your pull request or put a comment here, possibly with an @ mention of the developer that you want to consult.

navlalli committed:
Change names of orbital exponents

akohlmey commented:
@navlalli FYI, your code has whitespace issues (see the details of the `dev/pull_requests/checkstyle` test). To resolve those, please go to the `src` folder, type `make fix-whitespace` and commit the changes.

navlalli committed:
Fix whitespace

akohlmey committed:
Merge branch 'develop' into qtpie

akohlmey commented:
@navlalli FYI, I just merged your pull request branch with upstream because we are adding two more mandatory tests that need to be passed by all contributions. So before you make any more changes on your side, be certain to pull before your push. :wink:

navlalli committed:
Rename variables and function for min exponent

navlalli committed:
Add water examples featuring fix qtpie/reaxff

navlalli committed:
Remove unused code

navlalli committed:
Update author contributions

navlalli committed:
Remove additional fix name

navlalli committed:
Remove unnecessary tests

navlalli committed:
Remove unused header file

navlalli committed:
Change qeq to qtpie

navlalli committed:
Add fix qtpie/reaxff documentation

navlalli committed:
Fix whitespace

navlalli committed:
Add log files for qtpie/reaxff examples

navlalli committed:
Allow for output of effective electronegativities

navlalli committed:
Expand documentation

navlalli committed:
Fix whitespace

navlalli committed:
Make signs consistent with efield = -grad(phi)

navlalli committed:
Fix sign used for atom-style potential

A positive sign in front of the electric potential is consistent with
E = -grad(electric potential).

stanmoore1 commented:
@navlalli is this PR ready for review or are you still adding more commits?

navlalli commented:
@stanmoore1 Ready for review!

stanmoore1 commented on doc/src/fix_qtpie_reaxff.rst:
```diff
@@ -0,0 +1,196 @@
+
+This fix does not correctly handle interactions involving multiple
+periodic images of the same atom.  Hence, it should not be used for
+periodic cell dimensions less than 10 Angstroms.
```
Wouldn't this depend on the cutoff? Could say something like "Hence, it should not be used for
periodic cell dimensions less than the pairwise cutoff, typically 10 Angstroms."

navlalli replied:
Good point. I have updated the documentation (I have also added this change to the documentation of fix qeq/reaxff and fix acks2/reaxff).

stanmoore1 commented on doc/src/fix_qtpie_reaxff.rst:
```diff
@@ -0,0 +1,196 @@
+Related commands
+""""""""""""""""
+
+:doc:`pair_style reaxff <pair_reaxff>`, :doc:`fix qeq/reaxff <fix_qeq_reaxff>`
```
Need to list `fix acks2/reaxff` as a related command too

stanmoore1 commented on src/REAXFF/fix_qeq_reaxff.cpp:
```diff
@@ -1158,7 +1158,7 @@ void FixQEqReaxFF::get_chi_field()
       if (mask[i] & efgroupbit) {
         if (region && !region->match(x[i][0],x[i][1],x[i][2])) continue;
-        chi_field[i] = -efield->efield[i][3];
+        chi_field[i] = efield->efield[i][3];
```
Does this need to be propagated to other styles like ACKS2 and KOKKOS?

navlalli replied:
It seems that acks2/reaxff and the KOKKOS versions of qeq/reaxff and acks2/reaxff call `get_chi_field()` from `FixQEqReaxFF`, so I don't think any further changes are needed.

stanmoore1 reviewed (changes requested):
@navlalli overall this looks really good, I had a few comments/questions

navlalli committed:
Clarify restriction on periodic cell dimensions

navlalli commented:
@stanmoore1 Thank you for the review! I believe I have addressed the points. Please let me know if there is anything else.

jtclemm commented on src/REAXFF/fix_qtpie_reaxff.cpp:
```diff
@@ -0,0 +1,1231 @@
+    for (m = 0; m < n; m++) buf[m] = t[list[m]];
+  else if (pack_flag == 4)
+    for (m = 0; m < n; m++) buf[m] = atom->q[list[m]];
+  else if (pack_flag == 5) {
```
Is this option ever used? If so, the *comm_forward = 1* assignment in the constructor is insufficient. Ditto for reverse comm.

navlalli replied:
This is a good point. The `pack_flag = 5` option is never used (it is only used in `fix qeq/reaxff/omp`) and has now been removed.

jtclemm commented on src/REAXFF/fix_qtpie_reaxff.cpp:
```diff
@@ -0,0 +1,1231 @@
+
+  if (utils::strmatch(arg,"^reaxff")) {
+    reaxflag = 1;
+    Pair *pair = force->pair_match("^reaxff",0);
```
I understand this is called in the post constructor. If so, does this fix's behavior change if I defined a reaxff pair style after this fix is defined in the input script?

navlalli replied:
Correct. One gets `ERROR: No reaxff pair style for fix qtpie/reaxff (src/REAXFF/fix_qtpie_reaxff.cpp:243)` if you put `fix qtpie/reaxff` before defining the reaxff pair style. This behaviour is in alignment with `fix qeq/reaxff` and therefore `fix acks2/reaxff`.

navlalli committed:
Remove unused pack_flag = 5 options

navlalli committed:
Add fix qtpie/reaxff to pair_style reaxff docs

navlalli committed:
Fix whitespace

stanmoore1 committed:
Add more related commands to docs

stanmoore1 committed:
Merge branch 'develop' into qtpie

stanmoore1 reviewed (approved):
I approve
